### PR TITLE
feat: playground

### PR DESCRIPTION
**Pull request by mfw78**
# Description

This PR provides a full stack for running the CoW Protocol Infrastructure.

# Changes

- [x] Optimise `Dockerfile.binary` for caching
- [x] Implement CoW Protocol Playground

## How to test

1. [Follow the instructions](./playground/README.md).

## Dependencies

- [ ] Add blurb in CoW Protocol docs pointing to the playground

### CODE REVIEW

**Review by fleupold — COMMENTED** [review]
Comenting local changes I made so far (can also push them), but still running into some issues around starting the frontend (yarn i18n is failing)

**Comment by fedgiac** [comment]
Wow, that's very nice. I was able to follow the instructions to a point where I could interact with the frontend and the frontend could talk to the services. I wasn't able to get an order matched, probably because of the RPC I used:
1. MEV blocker would time out and cause all frontend's price requests to fail. The logs of the services showed a lot of timeouts. I was able to wrap ETH however.
2. https://ethereum.publicnode.com would show me prices from the frontend, but I wasn't able to execute any transaction on the node. At some point I started seeing a lot of timeouts anyway. 

So running this requires a good node.

By the way, I used Metamask since Rabby isn't available on Firefox. I discovered that apparently you can't add add a new network with chainId 1 from the Metamask interface, it forces you to use Infura.

<details><summary>This is how I managed to add the test network on Metamask.</summary>

Open your browser's dev console on any nonempty page and paste in the following snippet:

```js
await window.ethereum.request({
  method: 'wallet_addEthereumChain',
  params: [
    {
      chainId: '0x1',
      chainName: 'Local Network (Mainnet)',
      rpcUrls: ['http://localhost:8545'],
      nativeCurrency: {
        name: "Ethereum",
        symbol: "ETH",
        decimals: 18,
      },
    },
  ],
});
```
</details>

File: playground/README.md
Comment (fedgiac) [inline playground/README.md:13]: Something nice to write is how to clear up all state that has been accumulated by Docker so you can try out a new network, since I believe by default the database data is kept. Just a mention to `docker-compose -f docker-compose.fork.yml down --remove-orphans --volumes` would probably be enough.

Comment (mfw78) [reply playground/README.md:13]: Suggestions are happily merged :grin:

**Comment by mfw78** [comment]
Woah, good work on checking out those public RPC providers. I didn't even try :laughing:

> So running this requires a good node.

Most definitely this requires a solid node. Using CoW Protocol internal RPCs, @fleupold was able to settle a transaction on `mainnet`. I'd suggest that this is a great use case for those that are running a dappnode at home and able to tap an RPC.

I haven't tried with an xdai services configuration yet (I did have a bit of an attempt, but got stuck with no liquidity when fetching quotes). The standard wrapping etc worked.

**Comment by fedgiac** [comment]
I tried again with our staging mainnet node and I was able to wrap ETH, approve WETH, and settle an order with them!
There were still some timeouts going on every now and then. For example, the first time I visited the frontend page (as well as when I restarted the browser on the same page) it told me it couldn't show me the balances; also, there were timeouts in the services components. I suspect this is more because of Anvil than the node itself. Just waiting a bit (say, one minute) made the balance appear and I could trade as usual.

In case it helps others when testing (but not an issue with the PR): another thing that caused me a few issues is that Metamask was remembering that I sent a transaction in a previous test and was creating a transaction with a nonce too high, and so I was waiting for a confirmation of a transaction that couldn't be executed. Once realized that, it was easy to follow the instruction to [reset your account activity](https://support.metamask.io/hc/en-us/articles/360015488891-How-to-clear-your-account-activity-reset-account), but at first it wasn't clear this was happening.

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:30]: Needed to add this to be able to use `http://host.docker.internal:8545` to reach my `reth` node running on localhost.
```
    extra_hosts:
      - "host.docker.internal:host-gateway"
```

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:7]: I had to manually set a port that was **not** 8545 because `reth` was already listening on that port. Basically add this and replaced `8545` everywhere.
```suggestion
    command: --fork-url ${ETH_RPC_URL} --block-time 12 --port 8546
```

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:150]: Could not bind port `9001` because `reth` uses that by default.

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:134]: Could not use port `9000` because `lighthouse` uses that by default.

File: playground/.env.example
Comment (MartinquaXD) [inline playground/.env.example:1]: Would be nice to mention `http://host.docker.internal:8545` here if you are running a node locally.

**Review by MartinquaXD — COMMENTED** [review]
Awesome work!
Didn't test too extensively because I had to figure out a few adjustements because I was running `reth` locally as my `RPC`.
Other than that it was a bit strange that the frontend didn't notice that my transactions got mined (although metamask noticed).
To get my trade through I had to:
1. Wrap ETH
2. Reload page
3. Approve Contract
4. Reload
5. Trade
For some reason the trade event was correctly detected by the frontend. :thinking:

**Comment by MartinquaXD** [comment]
@mfw78 even if this might not be 100% perfect yet do you think we should just merge it for now? Doesn't really influence any other code and at least for people with a good node it already works.

File: playground/docker-compose.fork.yml
Comment (mfw78) [reply playground/docker-compose.fork.yml:30]: I would suggest when running a local node, such as `reth`, to bind it to `0.0.0.0`, and then add  an IP alias for your local NIC (or even just a "fake" bridge adaptor or similar), in which case then when running docker-in-docker, you will still be able to connect to this IP address (with `reth` on it), but avoid all the loopback catches when running docker from different contexts.

File: playground/docker-compose.fork.yml
Comment (mfw78) [reply playground/docker-compose.fork.yml:7]: As above. Suggest to run this within docker-in-docker (ie. from within the devcontainer). This ensures that it is agnostic to services being run locally.

File: playground/docker-compose.fork.yml
Comment (mfw78) [reply playground/docker-compose.fork.yml:150]: As above

File: playground/docker-compose.fork.yml
Comment (mfw78) [reply playground/docker-compose.fork.yml:134]: As above

File: playground/.env.example
Comment (mfw78) [reply playground/.env.example:1]: I think it's better to discourage this as per the above.

**Review by MartinquaXD — APPROVED** [review]
Approving as this does not influence any other code and it's already quite nice. Let's merge soon and iterate on the solution instead of never merging because it's not 100% perfect, yet.

**Comment by mfw78** [comment]
@MartinquaXD Would you mind one more review kindly before merge, given the changes in the docker generation?

**Review by MartinquaXD — COMMENTED** [review]
Forgot to submit my review amidst the troubles last week.

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:117]:
```suggestion
```

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:109]:
```suggestion
```

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:84]:
```suggestion
```

File: playground/docker-compose.fork.yml
Comment (MartinquaXD) [inline playground/docker-compose.fork.yml:74]:
```suggestion
```

**Review by MartinquaXD — COMMENTED** [review]
Some of the env variables have been removed by now.

File: playground/docker-compose.fork.yml
Comment (mfw78) [reply playground/docker-compose.fork.yml:117]: :laughing: This one...

File: Dockerfile
Comment (MartinquaXD) [reply Dockerfile:54]: @mfw78 you could build a container from this PR and try to use it in the infra repo to get the (probably) required infra change ready before merging this PR. Otherwise our staging pods will start crash looping.

**Pull request merged**
